src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java

```diff
@@ -97,20 +117,29 @@ private JsonArray formatBatch(String topic) {
             }
 
             JsonElement recordJSON = recordToJSON(record);
-            JsonObject message = populateMetadata(topic, recordJSON, record.timestamp());
+            JsonObject message = populateMetadata(topic, recordJSON, record.timestamp(), () -> kafkaHeadersToJsonElement(record));
             batchRecords.add(message);
         }
 
         return batchRecords;
     }
 
+    private JsonElement kafkaHeadersToJsonElement(SinkRecord sinkRecord) {
+        Map<String, Object> headerMap = stream(sinkRecord.headers().spliterator(), false)
+                .collect(toMap(Header::key, Header::value));
+
+        String jsonString = gson.toJson(headerMap);
+
+        return gson.fromJson(jsonString, JsonElement.class);
+    }
+
     private JsonElement recordToJSON(SinkRecord record) {
         byte[] rawJSONPayload = jsonConverter.fromConnectData(record.topic(), record.valueSchema(), record.value());
         String jsonPayload = new String(rawJSONPayload, StandardCharsets.UTF_8);
-        return new Gson().fromJson(jsonPayload, JsonElement.class);
+        return gson.fromJson(jsonPayload, JsonElement.class);
```

Comment:
Are we confident `gson` is completely stateless? It might looks like a detail, but if we're not 100% sure of this, let's maybe keep this unrelated fix for a separate PR.

Reply:
It looks like Gson should be completely stateless, but I saw some issues with it not long ago. Let's keep it the old way then!

Reply:
Reverted.